compiler/fir/raw-fir/fir-common/src/org/jetbrains/kotlin/fir/builder/BaseFirBuilder.kt

```diff
@@ -640,26 +639,40 @@ abstract class BaseFirBuilder<T>(val baseSession: FirSession, val context: Conte
         private val packageFqName: FqName,
         private val classFqName: FqName,
     ) {
+        private val firSource = source.toFirSourceElement()
+
+        private val copyName = Name.identifier("copy")
+        private val hashCodeName = Name.identifier("hashCode")
+        private val toStringName = Name.identifier("toString")
 
         fun generateMembers() {
             generateComponentFunctions()
             generateCopyFunction()
-            // TODO: equals, hashCode, toString
+            // TODO: need to check if these are overridable, e.g., user-given function, `final` one in hierarchy.
```

Comment:
Three regressions indicate that such condition should be checked, and synthetic ones should not be added if, e.g., user defines their own hashCode, or super type has its own with `final` modifier. Is this a right place to check and/or class builder has enough information to check that?

Reply:
Now I found how `psi2ir` avoids creating synthetic members if not necessary (or, in other words, how it creates those synthetic members only if needed): https://github.com/JetBrains/kotlin/blob/master/compiler/backend-common/src/org/jetbrains/kotlin/backend/common/FunctionsFromAnyGenerator.kt#L38 which goes to here: https://github.com/JetBrains/kotlin/blob/master/compiler/backend-common/src/org/jetbrains/kotlin/backend/common/CodegenUtil.kt#L139

Again, member scope from class descriptor and then `getContributedFunctions`, which are not available in wrapped descriptor. @mglukhikh any suggestions to do a similar checking? Note that I left a similar TODO in https://github.com/JetBrains/kotlin/pull/3286, resulting in very similar regression cases.

Reply:
I found my own answer: https://github.com/JetBrains/kotlin/pull/3294/commits/97ac4031fefa2aafe2ffd0fda779fdc11bf2915c#diff-2ed519bfe85d8b96c7fbba4ba954e24c